<a href="https://colab.research.google.com/github/knandagiri/data-repo/blob/master/19June_Group4_M0_Mini_Project_02_Linear_Algebra_and_Calculus.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Advanced Certification Program in Computational Data Science
## A program by IISc and TalentSprint
### Mini Project Notebook 2: Linear Algebra and Calculus
(Ungraded Mini-Project)

## Problem Statement

 The task is to advise a petroleum company on how to meet the demands of their customers for motor oil, diesel oil and gasoline.

## Learning Objectives

At the end of the experiment, you will be able to

* create arrays and matrices in python
* understand the concepts of linear equations
* solve the system of linear equations

### Data

From a barrel of crude oil, in one day, factory $A$ can produce 
* 20 gallons of motor oil, 
* 10 gallons of diesel oil, and 
* 5 gallons of gasoline 

Similarly, factory $B$ can produce 
* 4 gallons of motor oil, 
* 14 gallons of diesel oil, and 
* 5 gallons of gasoline
 
while factory $C$ can produce 
* 4 gallons of motor oil,
* 5 gallons of diesel oil, and 
* 12 gallons of gasoline 

There is also waste in the form of paraffin, among other things. Factory $A$ has 3 gallons of paraffin to dispose of per barrel of crude, factory $B$ 5 gallons, and factory $C$ 2 gallons.

**Note:** Your conclusion should include a discussion of the nature of the terms *unique*, *no solution*, *overdetermined* and *underdetermined* as they apply in the context of the oil plants.

In [ ]:
import numpy as np
import math as mth
import sympy as sy

### Create an array 

Create an array of size 2x3 with arbitrary values.

In [ ]:
# YOUR CODE HERE

In [ ]:
array_arbit = np.array(np.random.rand(6)).reshape(2,3)
array_arbit

array([[0.42246679, 0.43281865, 0.87680222],
       [0.77857767, 0.11676365, 0.98682828]])

### Create the system of Linear Equation

Suppose the current daily demand from distributors is 6600 gallons of motor oil, 5100 gallons of diesel oil and 3100 of gasoline.

Set up the system of equations which describes the above situation. Please include the units as well.

Let the number of barrels used by factory $A$, $B$ and $C$ are $x$, $y$ and $z$ respectively.

Then the system of linear equations will be

$$Motor\ oil:\ \ \ 20x + 4y + 4z = 6600$$

$$Diesel\ oil:\ \ \ 10x + 14y + 5z = 5100$$

$$Gasoline:\ \ \ 5x + 5y + 12z = 3100$$

### Solve the system of Linear Equation

How many barrels of crude oil each plant should get in order to meet the demand as a group. Remember that we can only provide each plant with an integral number of barrels.

In [ ]:
# YOUR CODE HERE
coef_matrix = np.array([20, 4, 4,10, 14, 5, 5, 5, 12]).reshape(3,3)
coef_matrix
value_matrix = np.array([6600, 5100, 3100])
type(value_matrix)
# LEARNINGS 
# matrix vs array
# np.array([6600, 5100, 3100]) w and w/o reshape - CHECK shape of array
# np.asarray vs np.array

numpy.ndarray

In [ ]:
# defining the coefficient matrix
# coef_matrix = np.matrix('20, 4, 4; 10, 14, 5; 5, 5, 12')
coef_matrix = np.array([[20, 4, 4],[10, 14, 5], [5, 5, 12]])

# defining the constant value matrix
# value_matrix = np.array('6600, 5100, 3100').reshape(3,1)
value_matrix = np.array([[6600],[5100], [3100]])

# use np.linalg.solve to solve for x
X = np.linalg.solve(coef_matrix, value_matrix)

# displaying the value of x, y and z
#print(X)

# No has to be an integer
Xno = np.ceil(X)

print( X,"No. of barrels for Company A, B and C : \n", Xno)


#****************************************#
# Ceiling is pushing 85.00 to 86

# Rounding off problem
#Xno = np.asmatrix(X)
#Xno.put( [0, 1, 2], [np.ceil(X[i]) if mth.modf(X[i])[0] > 0.0 else np.floor(X[i]) for i in range(X.size)] )
#print(mth.modf(X[0]))

[[287.25]
 [128.75]
 [ 85.  ]] No. of barrels for Company A, B and C : 
 [[288.]
 [129.]
 [ 85.]]


Suppose the total demand for all products **doubled**. What would the solution now be? How does it compare to the original solution? Why, mathematically, should this have been expected?

In [ ]:
# YOUR CODE HERE

In [ ]:
Motor oil:   20x+4y+4z=6600 * 2 

Diesel oil:   10x+14y+5z=5100 * 2

Gasoline:   5x+5y+12z=3100 * 2

In [ ]:
# Column Matrix X : 3 by 1 gives me the values for the earlier demand. Scale it by 2 to meet the new demand

print("As per double demand, requirement of No. of Barrels for Company A, B and C = \n", np.ceil(2 * X))


As per double demand, requirement of No. of Barrels for Company A, B and C = 
 [[575.]
 [258.]
 [170.]]


In [ ]:
# Solving the equation fully again with Scaled Demand 
coefMatrix = np.matrix( [ [20, 4, 4], [10, 14, 5], [5, 5, 12] ])
valueMatrix = np.matrix([13200, 10200, 6200]).reshape(3,1)

# 1. Solving Linear System equation with linalg.solve()
solnMatrix = np.linalg.solve(coefMatrix, valueMatrix)

print("\nFor 2 times demand, no of barrels = \n", np.ceil(solnMatrix))

# 2. Linear System equation solved using Inverse => [A]X = [B] => X = [A]^-1 . [B]
#np.ceil( np.dot(np.linalg.inv(coefMatrix), valueMatrix) )

# When I use Inverse to solve equations, values are same as linalg.solve() but when applying np.ceil(), the value without decimal point is jumping UP.


For 2 times demand, no of barrels = 
 [[575.]
 [258.]
 [170.]]


In [ ]:
factor = np.int64(np.round(np.divide(np.ceil(solnMatrix),Xno)))
print(np.gcd.reduce(factor))

[[2]]


Suppose that the company acquires another group of distributors and that the daily demand of this group is 2000 gallons of motor oil, 4000 gallons of gasoline, and 4000 gallons of diesel oil. How would you set up production of just this supply? Are there any options (more than one way)?

In [ ]:
# YOUR CODE HERE

In [ ]:
Motor oil:   20x+4y+4z=2000 

Diesel oil:   10x+14y+5z=4000

Gasoline:   5x+5y+12z=4000

In [ ]:
# Find Inverse of A
Acoef = np.matrix('20, 4, 4; 10, 14, 5; 5, 5, 12')
Aval = np.matrix('2000, 4000, 4000').reshape(3,1)

Y = np.linalg.solve(Acoef, Aval)

# No has to be an integer
Yno = np.ceil(Y)
print("\nNo of barrels required for second distributer : \n", Yno)


#**************************************#
#For inverse np.linalg.solve(Ainverse, Aval) WILL NOT work. Dot product of Ainverse & Values needs to be taken
#inv_Acoef = np.linalg.inv(Acoef)
#Y = np.dot(inv_Acoef, Aval)
#print(Y)


No of barrels required for second distributer : 
 [[ 13.]
 [188.]
 [250.]]


Next, calculate the needs of each factory (in barrels of crude, as usual) to meet the total demand of both groups of distributors. When you have done this, compare your answer to results already obtained. What mathematical conclusion can you draw?

In [ ]:
# YOUR CODE HERE

In [ ]:
# To meet the Total demand of Doubled demand and the current demand : Matrix Addition 
print("\nEarlier demand of production -  X (from first question) : \n", X)
print("Current demand of new distributor - Y ( from the previous question) : \n", Y)
Total_matrix =  X + Y
print("\nTotal demand of barrels is X + Y = \n", np.ceil(Total_matrix) )


Earlier demand of production -  X (from first question) : 
 [[287.25]
 [128.75]
 [ 85.  ]]
Current demand of new distributor - Y ( from the previous question) : 
 [[ 12.5]
 [187.5]
 [250. ]]

Total demand of barrels is X + Y = 
 [[300.]
 [317.]
 [335.]]


In [ ]:
# Find Inverse of A
Acoef = np.matrix('20, 4, 4; 10, 14, 5; 5, 5, 12')
Aval = np.matrix('8600, 9100, 7100').reshape(3,1)

Y = np.linalg.solve(Acoef, Aval)

# No has to be an integer
Yno = np.ceil(Y)
print("\nNo of barrels required for second distributer : \n", Yno)


No of barrels required for second distributer : 
 [[300.]
 [317.]
 [335.]]


In [ ]:
import numpy.matlib as mat

#Wastage / barrel for all 3 Companies - A : 3 gallons , B : 5 gallons and C : 2 gallons

# Creating Empty matrix with shape 3 by 1 and then putting in values - 3, 5, 2 . Should be a COLUMN MATRIX :: 3 by 1
waste_matrix = mat.empty((3,1))
waste_matrix.put([0, 1, 2], [3, 5, 2])    # [0, 1, 2] are the indices and the next list is the values

print("Waste : \n", waste_matrix)
print("No. of barrels / company : \n", np.ceil(Total_matrix) )

# Element wise multiplication required so np.array provides np.multiply and A * B for element wise multiplication
print("\nTotal waste production company-wise : \n", np.array(waste_matrix) * np.array(np.ceil(Total_matrix)) )

Waste : 
 [[3.]
 [5.]
 [2.]]
No. of barrels / company : 
 [[300.]
 [317.]
 [335.]]

Total waste production company-wise : 
 [[ 900.]
 [1585.]
 [ 670.]]


### Sensitivity and Robustness

In real life applications, constants are rarely ever exactly equal to their stated value; certain amounts of uncertainty are always present. This is part of the reason for the science of statistics. In the above model, the daily productions for the plants would be averages over a period of time. Explore what effect small changes in the parameters have on the output.

To do this, pick any 3 coefficients, one at a time, and increase or decrease them by 3%. For each case , note what effect this has on the solution, as a percentage change. Can you draw any overall conclusion?

In [ ]:
# YOUR CODE HERE

In [ ]:
Motor oil:   20x+4y+4z=6600 

Diesel oil:   10x+14y+5z=5100 

Gasoline:   5x+5y+12z=3100

In [ ]:
# 3 % increase means Multiplying with a scalar 1.03
coef_matrix1 = np.array([20*1.03, 4, 4,10*1.03, 14, 5, 5*1.03, 5, 12]).reshape(3,3)
coef_matrix2 = np.array([20*0.97, 4, 4,10*0.97, 14, 5, 5*0.97, 5, 12]).reshape(3,3)

# defining the constant value matrix
value_matrix = np.matrix('6600, 5100, 3100').reshape(3,1)

# Solve linear equation
soln_mat = np.linalg.solve(coef_matrix, value_matrix) # No change 
soln_mat1 = np.linalg.solve(coef_matrix1, value_matrix) # 3% increase for A
soln_mat2 = np.linalg.solve(coef_matrix2, value_matrix) # 3% decrease for A

print(soln_mat[0],soln_mat1[0],soln_mat2[0])
print(soln_mat1[0] - soln_mat[0],soln_mat2[0] - soln_mat[0])

# ******************************************************#

# Rounding off problem for float decimal values
#tmp = np.asmatrix(soln_mat)
#tmp.put( [0, 1, 2], [np.ceil(soln_mat[i]) if mth.modf(soln_mat[i])[0] > 0.0 else np.floor(soln_mat[i]) for i in range(soln_mat.size)] )

#print("\n Barrels required : \n", tmp )

[[287.25]] [[278.88349515]] [[296.13402062]]
[[-8.36650485]] [[8.88402062]]


In [ ]:
import itertools

perm = np.asarray(list(itertools.permutations(range(3),2)))
np.random.shuffle(perm)
perm = perm[:3]
print(perm)

new_coeff_mat = np.asarray(coef_matrix.copy(), dtype= np.float64)
pctChange = 0.03

for i in perm:
  sign = np.array([-1, 1])
  np.random.shuffle(sign)
  new_coeff_mat[i[0],i[1]] = new_coeff_mat[i[0],i[1]] + sign[0] * 0.03

print("\noriginal coeff mat")
print(coef_matrix)

print("\nNew coeff mat")
print(new_coeff_mat)

new_barrel_vec = np.linalg.solve(new_coeff_mat,value_matrix)
new_barrel_vec_int = np.int64(np.ceil(new_barrel_vec))
print(new_barrel_vec_int)

change = [(new_barrel_vec[i] - X[i]) / 100.0 for i in range(len(X))]
print("\nchange: " + str(change))

[[0 1]
 [1 0]
 [1 2]]

original coeff mat
[[20  4  4]
 [10 14  5]
 [ 5  5 12]]

New coeff mat
[[20.    3.97  4.  ]
 [10.03 14.    5.03]
 [ 5.    5.   12.  ]]
[[288]
 [128]
 [ 86]]

change: [array([0.00358569]), array([-0.01177104]), array([0.00341056])]


### A Plant Off-Line

Suppose factory $C$ is shut down by the EPA (Environmental Protection Agency) temporarily for excessive emissions into the atmosphere. If your demand is as it was originally (6600, 5100, 3100), what would you now say about the companies ability to meet it? What do you recommend they schedule for production now?

In [ ]:
# YOUR CODE HERE

In [ ]:
import sympy as sy

In [ ]:
Motor oil:   20x+4y+0z=6600 

Diesel oil:   10x+14y+0z=5100 

Gasoline:   5x+5y+0z=3100

In [ ]:
# 1.
# Using the Reduced Row Echelon Form of Sympy :: rref
M = sy.Matrix( [ [20, 4, 0, 6600], [10, 14, 0, 5100], [5, 5, 0, 3100] ])
aug_M = M.rref()
aug_M

(Matrix([
 [1, 0, 0, 0],
 [0, 1, 0, 0],
 [0, 0, 0, 1]]), (0, 1, 3))

In [ ]:
print("\n The rref of the matrix shows x = 0 and y = 0 and z does not exist !!! ")
#  this is a unique solution X = 19, Y = 11 and Z = 11
 [1, 0, 0, 19],
 [0, 1, 0, 11],
 [0, 0, 1, 11]
# No Solution. as 0 != 1 in the last row
 [1, 0, 0, 0],
 [0, 1, 0, 0],
 [0, 0, 0, 1]
# Many solutions example
 [1, 0, 0, 10],
 [0, 1, 1, 20],
 [0, 1, 1, 156]


 The rref of the matrix shows x = 0 and y = 0 and z does not exist !!! 


In [ ]:
# 2. Solving Equation 1 & 2 :
# x and y could be solved with any 2 equations 
M = sy.Matrix( [[20, 4, 6600] , [10, 14, 5100]] )
Soln = M.rref()
Soln
print("\n X & Y :: ", Soln[0].col(-1) )

#2. Solving Equation 2 & 3 :
# x and y could be solved with any 2 equations 
M = sy.Matrix( [ [10, 14, 5100] , [5, 5, 3100] ] )
Soln = M.rref()
Soln
print("\n X & Y :: ", Soln[0].col(-1) )

#2. Solving Equation 1 & 3 :
# x and y could be solved with any 2 equations 
M = sy.Matrix( [ [20, 4, 6600] , [5, 5, 3100] ] )
Soln = M.rref()
Soln
print("\n X & Y :: ", Soln[0].col(-1) )

print("\n All 3 set of equation gave 3 different values for X and Y :: NOT A DEFINED SOLUTION !!! ")


 X & Y ::  Matrix([[300], [150]])

 X & Y ::  Matrix([[895], [-275]])

 X & Y ::  Matrix([[515/2], [725/2]])

 All 3 set of equation gave 3 different values for X and Y :: NOT A DEFINED SOLUTION !!! 


In [ ]:
#3. Least Square Method 

# defining the coefficient matrix
coef_matrix = np.matrix('20, 4, 0; 10, 14, 0; 5, 5, 0')

# defining the constant value matrix
value_matrix = np.matrix('6600, 5100, 3100').reshape(3,1)

# use np.linalg.solve to solve for x
X = np.linalg.lstsq(coef_matrix, value_matrix, rcond = -1)

# displaying the value of x, y and z
print(X)

(matrix([[299.47204969],
        [168.47826087],
        [  0.        ]]), matrix([], shape=(1, 0), dtype=float64), 2, array([25.79117677,  9.83947158,  0.        ]))


In [ ]:
#rcond : float, optional
#    Cut-off ratio for small singular values of a.
# Rank

### Buying another plant

This situation has caused enough concern that the CEO is considering buying another plant, identical to the third, and using it permanently. Assuming that all 4 plants are on line, what production do you recommend to meet the current demand (5000, 8500, 10000)? In general, what can you say about any increased flexibility that the 4th plant might provide?

Let the number of barrels used by factory $A$, $B$, $C$ and $D$ are $x$, $y$, $z$ and $w$ respectively.

Then the system of linear equations will be

$$20x + 4y + 4z + 4w = 5000$$

$$10x + 14y + 5z + 5w = 8500$$

$$5x + 5y + 12z + 12w = 10000$$

The above system of linear equation has fewer equations than variables, hence it is *underdetermined* and cannot have a unique solution. In this case, there are either infinitely many solutions or no exact solution. We can solve it by keeping $w$ as constant and using [rref](http://linear.ups.edu/html/section-RREF.html) form to solve the system of linear equation.

To know about rref implementation in python refer [here](https://docs.sympy.org/latest/tutorial/matrices.html#rref).

In [44]:
import sympy as sy

# create symbol 'w'
w = sy.Symbol("w")            
A_aug = sy.Matrix([[20, 4, 4, 5000-4*w], 
                   [10, 14, 5, 8500-5*w], 
                   [5, 5, 12, 10000-12*w]])
# show rref form
A_aug.rref()                  

(Matrix([
 [1, 0, 0,   195/4],
 [0, 1, 0,  1325/4],
 [0, 0, 1, 675 - w]]), (0, 1, 2))

From the above result, it can be seen that 4th plant will share the number of barrels required by the 3rd plant only, while the requirement of 1st and 2nd plant will remain unaffected.

### Calculate the amount of Paraffin supplied

The company has just found a candle company that will buy its paraffin. Under the current conditions (i.e, after buying another plant) for demand (5000, 8500, 10000), how much can be supplied to them per day?

According to the problem statement, factory $A$ has 3 gallons of paraffin to dispose of per barrel of crude oil, factory $B$ 5 gallons, and factory $C$ 2 gallons.

In [46]:
# YOUR CODE HERE
# z + w = 675
# 195/4 * 3 + 1325/4 * 5 + (675) * 2 (= z * 2 + w *2 = (z + w) *2 = 675 *2)
paraffin = sy.Matrix([3,5,2]).T
barrel_count = A_aug.rref()[0][:,3]
print(barrel_count)
paraffin_prod = paraffin * barrel_count
print("\nParaffin production: " + str(paraffin_prod[0]))
print("The answer is 195/4 * 3 + 1325/4 * 5 + (675) * 2")

Matrix([[195/4], [1325/4], [675 - w]])

Paraffin production: 6305/2 - 2*w
The answer is 195/4 * 3 + 1325/4 * 5 + (675) * 2


In [ ]:
import numpy.matlib as mat

#Wastage / barrel for all 3 Companies - A : 3 gallons , B : 5 gallons and C : 2 gallons

waste_matrix = mat.empty((3,1))
waste_matrix.put([0, 1, 2], [3, 5, 2])    # [0, 1, 2] are the indices and the next list is the values

# Taking A_aug from previous question
s = A_aug.rref()[0]
s = mat.matrix(s.col(-1)[0:2])
s = s.reshape(2,1,)

print("\nWaste : \n", waste_matrix)
print("\n Earlier soln for variables x and y from rref : \n", s)

# Element wise multiplication :: Array Multiplication
final = np.array(waste_matrix[0:2]) * np.array(s)
tmpmat = np.asmatrix(s)
tmpmat.put([0, 1], [np.ceil(i) for i in final] )
print("\n Supply of paraffin by company A and B : \n", tmpmat)


Waste : 
 [[3.]
 [5.]
 [2.]]

 Earlier soln for variables x and y from rref : 
 [[195/4]
 [1325/4]]

 Supply of paraffin by company A and B : 
 [[147]
 [1657]]


In [ ]:
Motor oil:   20x+4y+4z=5000 

Diesel oil:   10x+14y+5z=8500

Gasoline:   5x+5y+12z=10000

In [ ]:
#2. Solving full equation
import numpy.matlib as mat

Acoef = np.matrix('20, 4, 4; 10, 14, 5; 5, 5, 12')
Aval = np.matrix('5000, 8500, 10000').reshape(3,1)

Y = np.linalg.solve(Acoef, Aval)
Yno = np.ceil(Y)

# Creates an empty matrix 
waste_matrix = mat.empty((3,1))

# np.matrix has a function put to add values to a matrix based on the indices. First argument is a list of indices and second is the list of values
waste_matrix.put([0, 1, 2], [3, 5, 2])

print( np.array(Yno) * np.array(waste_matrix) )

[[ 147.]
 [1660.]
 [1350.]]


### Selling the first plant

The management is also considering selling the first plant due to aging equipment and high workman's compensation costs for the state it is located in. They would like to know what this would do to their production capability. Specifically, they would like an example of a demand they could not meet with only plants 2 and 3, and also what effect having plant 4 has (recall it is identical to plant 3). They would also like an example of a demand that they could meet with just plants 2 and 3. Any general statements you could make here would be helpful.

Let the number of barrels used by factory $B$, $C$ and $D$ are $y$, $z$ and $w$ respectively.

When considering only plants 2 and 3, and demand (5000, 8500, 10000) then we have

$$4y + 4z = 5000$$

$$14y + 5z = 8500$$

$$5y + 12z = 10000$$

In [ ]:
# YOUR CODE HERE
4𝑦+4𝑧+4w=5000 
14𝑦+5𝑧+5w=8500
5𝑦+12𝑧+12w=10000

In [ ]:
import sympy as sy

In [ ]:
# 1.
# Solving every pair of equations
# Solving Equation 1 and 2 
M = sy.Matrix( [[4, 4, 5000],[14, 5, 8500]] )
Soln = M.rref()
#print(Soln)
print("\n No of barrels required by Company 2 and 3 : \n", Soln[0].col(-1))

# Solving Equation 2 and 3 
M = sy.Matrix( [[14, 5, 8500],[5, 12, 10000]] )
Soln = M.rref()
#print("\n", Soln)
print("\n No of barrels required by Company 2 and 3 : \n", Soln[0].col(-1))

# Solving Equation 1 and 3 
M = sy.Matrix( [[4, 4, 5000],[5, 12, 10000]] )
Soln = M.rref()
#print("\n", Soln)
print("\n No of barrels required by Company 2 and 3 : \n", Soln[0].col(-1))

print("\nMANY solutions for y and z :: NO defined solution !!!")


 No of barrels required by Company 2 and 3 : 
 Matrix([[250], [1000]])

 No of barrels required by Company 2 and 3 : 
 Matrix([[4000/11], [7500/11]])

 No of barrels required by Company 2 and 3 : 
 Matrix([[5000/7], [3750/7]])

MANY solutions for y and z :: NO defined solution !!!


In [ ]:
# 2.
# Least Square Method
# defining the coefficient matrix
coef_matrix = np.matrix('4, 4; 14, 5; 5, 12')

# defining the constant value matrix
value_matrix = np.matrix('5000, 8500, 10000').reshape(3,1)

# use np.linalg.solve to solve for x
X = np.linalg.lstsq(coef_matrix, value_matrix, -1)

# displaying the value of x, y and z
print(X)

(matrix([[369.30178881],
        [695.03750721]]), matrix([[607616.84939411]]), 2, array([18.95513121,  7.91852263]))


Taking 4th plant into consideration.
Let the number of barrels used by factory $B$, $C$ and $D$ are $y$, $z$ and $w$ respectively.

Then for demand (5000, 8500, 10000) the system of linear equations will be

$$4y + 4z + 4w = 5000$$

$$14y + 5z + 5w = 8500$$

$$5y + 12z + 12w = 10000$$

Solve it using rref form.

In [ ]:
# YOUR CODE HERE

In [47]:
# Equation could be solved as 3 variables and 3 equations
# sy is sympy which has been imported earlier

M = sy.Matrix( [ [4, 4, 4, 5000], [14, 5, 5, 8500], [5, 12, 12, 10000] ] )
aug_rref = M.rref()

print(aug_rref)

# Equations :
# y = 0
# z + w = 0
print("\n No Solution : Demand - Supply will NOT be met !!!")

(Matrix([
[1, 0, 0, 0],
[0, 1, 1, 0],
[0, 0, 0, 1]]), (0, 1, 3))

 No Solution : Demand - Supply will NOT be met !!!


Now, changing demand to (6600, 5100, 3100) and solving the system of equation using rref form.

In [ ]:
# YOUR CODE HERE

In [ ]:
# Equation could be solved as 3 variables and 3 equations
M = sy.Matrix( [ [4, 4, 4, 6600], [14, 5, 5, 5100], [5, 12, 12, 3100] ] )
aug_rref = M.rref()

print(aug_rref)

# Equations :
# y = 0
# z + w = 0
print("\n No Solution : Demand - Supply will NOT be met !!!")

(Matrix([
[1, 0, 0, 0],
[0, 1, 1, 0],
[0, 0, 0, 1]]), (0, 1, 3))

 No Solution : Demand - Supply will NOT be met !!!


### Set rates for Products

Company wants to set the rates of motor oil, diesel oil, and gasoline. For this purpose they have few suggestions given as follows: 

* 100, 66, 102 Rupees per gallon,

* 104, 64, 100 Rupees per gallon,

* 102, 68, 98 Rupees per gallon, and

* 96, 68, 100 Rupees per gallon 

for motor oil, diesel oil, and gasoline respectively.

Using matrix multiplication, find the rates which results in maximum total price.

Let $M$ denotes the matrix such that rows represents different plants (A, B and C), columns represents different products (motor oil, diesel oil and gasoline) and each value represents production of that product from one barrel of crude oil for that plant.

$$M = \begin{bmatrix}
20 & 10 & 5 \\ 
4 & 14 & 5  \\ 
 4 & 5 & 12  
\end{bmatrix}$$

Also, $R$ is a matrix having different rates as its columns.

$$R = \begin{bmatrix}
100 & 104 & 102 & 96 \\ 
66 & 64 & 68 & 68  \\ 
102 & 100 & 98 & 100  
\end{bmatrix}$$

In [ ]:
# YOUR CODE HERE

In [49]:
# Matrix multiplication :- dot product
# sy is sympy which has been imported earlier

M = sy.Matrix( [ [20, 10, 5], [4, 14, 5], [4, 5, 12] ] )
R = sy.Matrix( [ [100, 104, 102, 96], [66, 64, 68, 68], [102, 100, 98, 100] ] )
TotCost = M * R
print("\n Rowwise : Company A, B and C : Total Production-Value :: \n", TotCost)
print("\nMax cost for Company A : ", max(TotCost.row(0)) )
print("\nMax cost for Company B : ", max(TotCost.row(1)) )
print("\nMax cost for Company C : ", max(TotCost.row(2)) )
print("\nMax cost for  price set 1: ", sum(TotCost.col(0)) )
print("\nMax cost for  price set 2: ", sum(TotCost.col(1)) )
print("\nMax cost for  price set 3: ", sum(TotCost.col(2)) ) # This the best process
print("\nMax cost for  price set 4: ", sum(TotCost.col(3)) )



 Rowwise : Company A, B and C : Total Production-Value :: 
 Matrix([[3170, 3220, 3210, 3100], [1834, 1812, 1850, 1836], [1954, 1936, 1924, 1924]])

Max cost for Company A :  3220

Max cost for Company B :  1850

Max cost for Company C :  1954

Max cost for  price set 1:  6958

Max cost for  price set 2:  6968

Max cost for  price set 3:  6984

Max cost for  price set 4:  6860


Max cost for  price set 3:  6984 - Best one

### Marginal Cost

The total cost $C(x)$ in Rupees, associated with the production of $x$ gallons of gasoline is given by 

$$C(x) = 0.005 x^3 – 0.02 x^2 + 30x + 5000$$

Find the marginal cost when $22$ gallons are produced, where, marginal cost means the instantaneous rate of change of total cost at any level of output.

In [ ]:
# YOUR CODE HERE

In [51]:
# d/dx
# sy is sympy which has been imported earlier

# defining the symbols used in the expression for C(x)
x = sy.symbols('x')

# defining the expression C(x)
Cx = 0.005 * x**3 - 0.02 * x**2 + 30 * x + 5000
costexpr = sy.Derivative(Cx, x).doit()

print("\nCost Expression after differentiation :: ", costexpr)

# Substituting value of x = 22
marginalCost = costexpr.subs(x, 22)
print("\n Marginal Cost for x at 22 = ", round(marginalCost, 2) )



Cost Expression after differentiation ::  0.015*x**2 - 0.04*x + 30

 Marginal Cost for x at 22 =  36.38


### Marginal Revenue

The total revenue in Rupees received from the sale of $x$ gallons of a motor oil is given by $$R(x) = 3x^2 + 36x + 5.$$ 

Find the marginal revenue, when $x = 28$, where, marginal revenue means the rate of change of total revenue with respect to the number of items sold at an instant.

In [ ]:
# YOUR CODE HERE

In [52]:
#d/dx
# sy is sympy which has been imported earlier

# Defining Rx as a function so no need to define symbols seperately
Rx = sy.Function('Rx')
Rx = 3 * x**2 + 36 * x + 5
revexpr = sy.diff(Rx, x) 

print("\n Revenue expression :: ", revexpr)

marginalRevenue = revexpr.evalf(subs = {x : 28})

marginalRevenue_2 = revexpr.evalf(subs = {x : 30})

print("\n Marginal Reveneue when x = 28  is ", round(marginalRevenue) )

print("\n Marginal Reveneue when x = 30  is ", round(marginalRevenue_2) )



 Revenue expression ::  6*x + 36

 Marginal Reveneue when x = 28  is  204

 Marginal Reveneue when x = 30  is  216


### Pour crude oil in tank

In a cylindrical tank of radius 10 meter, crude oil is being poured at the rate of 314 cubic meter per hour. Then find

* the rate at which the height of the tank is increasing, and
* the height of crude oil in tank after 2 hours.

In [ ]:
# YOUR CODE HERE

In [ ]:
# 1. Rate at which height of the tank is increasing
# Given :
r = 10
rate = 314  # m3/hr
# V = 3.14 * r**2 * h

# Define symbols

h = V / 3.14 * r**2   
dhdt = rate / (3.14 * r**2)   # dV/dt = rate of change

print("\n Rate at which height is increasing = ", dhdt, " m3/hr")

# Mathematically from Shashank's jpg :
# dh/dt = d/dt ( V/3.14 * r**2) 
#       = 1/3.14 x r**2 x (dV/dt)
#       = 1/3.14 x 10**2 x 314 m3/hour = 1 m3/hour 

# Height of tank after 2 hours
H = (2 * rate) / (3.14 * r**2) 

print("\n Height of tank after 2 hours : ", H, " m")


 Rate at which height is increasing =  1.0  m3/hr

 Height of tank after 2 hours :  2.0  m


In [53]:
# YOUR CODE HERE
# v = pi r**2 h 
# dv/dt = pi r**2 dh/dt
rate = 314 / (np.pi * 10 * 10)
print(str(rate) + " m/h" )
print("After 2 hours : " + str(2*rate) + " m" )

0.9994930426171027 m/h
After 2 hours : 1.9989860852342054 m
